### PR TITLE
[#320] fix docker plugin config to account for change from orphedomos

### PR DESCRIPTION
**Pull request by ewilkins-csi**
Two main changes enacted though archetype/template updates and migrations:

- All docker modules have to include the docker plugin in their build
- Any POM modules that configure the docker plugin must do so in a pluginManagement section to avoid errors

### CODE REVIEW

File: foundation/foundation-mda/src/main/resources/templates/general-docker/inference.docker.pom.xml.vm
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/resources/templates/general-docker/inference.docker.pom.xml.vm:1]: I: Primarily whitespace changes. Only real change was adding the docker-maven-plugin.

File: foundation/foundation-mda/src/main/resources/templates/general-docker/it.java.docker.file.vm [hunk 1]
```diff
@@ -3,8 +3,7 @@
 # GENERATED DOCKERFILE - please ***DO*** modify.
 #
 # Generated from: ${templateName}
-ARG DOCKER_BASELINE_REPO_ID
-FROM ${DOCKER_BASELINE_REPO_ID}boozallen/openjdk:11-slim
+FROM openjdk:11-slim
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/resources/templates/general-docker/it.java.docker.file.vm:6]: I: This image is not available in our ghcr.io registry, and the docker.io image is functionally what we need anyway. The 1.7 release notes address fixing this on upgrade.

File: foundation/foundation-mda/src/main/resources/templates/general-docker/training.docker.pom.xml.vm
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/resources/templates/general-docker/training.docker.pom.xml.vm:1]: I: Same as before &mdash; primarily whitespace changes

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/AbstractPomMigration.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+public abstract class AbstractPomMigration extends AbstractAissembleMigration {
+    protected static final Logger logger = LoggerFactory.getLogger(AbstractPomMigration.class);
+
+    public static final String POM = "pom";
+    public static final String GUARANTEED_TAG = "<modelVersion>";
+
+    protected String indent;
+
+    protected void detectAndSetIndent(File file) {
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/AbstractPomMigration.java:31]: I: I think ideally we want to bake this into the POM utils in Baton, but I didn't want to upgrade the Baton version for a patch release. I'll open a PR in that repo for this later.

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_8_0/Fabric8LocationMigration.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        boolean removePluginsSection = removePluginFromPlugins(modifications, dockerPlugin, build);
+        addPluginToManagement(modifications, build, removePluginsSection, content);
+    }
+
+    private void addPluginToManagement(PomModifications modifications, BuildBase build, boolean removePluginsSection, List<String> content) {
+        InputLocation insertLocation;
+        PluginManagement management = build.getPluginManagement();
+        if (management == null ) {
+            InputLocation pluginsStart = build.getLocation("plugins" + START);
+            insertLocation = removePluginsSection ? build.getLocation(END) : pluginsStart;
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_8_0/Fabric8LocationMigration.java:86]: I: Just allows a tidier insertion by putting the new plugin management block above plugins if that section isn't being completely removed.

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/util/AissembleFileUtils.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    /**
+     * Copy lines between startLine and endLine, inclusive, from the file.
+     *
+     * @param file the file to copy lines from
+     * @param startLine the index of the first line to copy
+     * @param endLine the index of the last line to copy
+     * @return a list of the copied lines
+     */
+    public static List<String> getLines(Path file, int startLine, int endLine) {
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/util/AissembleFileUtils.java:30]: I: I _think_ this is the same as the indent detection for POMs and can/should be contributed back to Baton.

File: foundation/foundation-upgrade/src/test/java/com/boozallen/aissemble/upgrade/migration/v1_8_0/AiopsReferenceJavaMigrationSteps.java [hunk 5]
```diff
@@ -50,7 +50,7 @@ public void theAiopsReferenceJavaMigrationExecutes() {
 
     @Then("the references are updated to com.boozallen.aissemble.data.access")
     public void theReferencesAreUpdatedToComBoozallenAissembleDataAccess() {
-        validatedFile = getTestFile("/v1_8_0/AiopsReferenceJavaMigration/validation/DataAccessRecordValidated.java");
+        validatedFile = getTestFile("v1_8_0/AiopsReferenceJavaMigration/validation/DataAccessRecordValidated.java");
```
Comment (ewilkins-csi) [inline foundation/foundation-upgrade/src/test/java/com/boozallen/aissemble/upgrade/migration/v1_8_0/AiopsReferenceJavaMigrationSteps.java:53]: I: I originally didn't have the logic to strip leading file separators in `getTestFile`, reasoning that a leading slash means an absolute path and should be treated as one.  However, since the parameter is clearly labelled `subPath`, I decided to reverse course on that. Happy to revert these changes if anyone feels strongly.

File: foundation/foundation-upgrade/src/test/resources/specifications/v1_8_0/orphedomos-to-fabric8-migration.feature [hunk 6]
```diff
@@ -24,3 +24,55 @@ Feature: Migrate from using Orphedomos to Fabric8 docker-maven-plugin
...
+    When the add Fabric8 migration executes
+    Then the docker-maven-plugin is added to the build section with minimal configuration
+
+  Scenario: Add plugin, no build section
+    Given a POM with the "docker-build" packaging type
+    And there is no build section
+    When the add Fabric8 migration executes
+    Then the docker-maven-plugin is added to a new build section with minimal configuration
+
+  Scenario: Add plugin, no build section
```
Comment (csun-cpointe) [inline foundation/foundation-upgrade/src/test/resources/specifications/v1_8_0/orphedomos-to-fabric8-migration.feature:45]: A: maybe a different scenario name: skip migration if already in build section?

Comment (carter-cundiff) [inline foundation/foundation-upgrade/src/test/resources/specifications/v1_8_0/orphedomos-to-fabric8-migration.feature:45]: Think this scenario outline needs to be updated, seems to be a copy from above

File: foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_8_0/Fabric8LocationMigration.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+ * pluginManagement section.
+ */
+public class Fabric8LocationMigration extends AbstractPomMigration {
+    @Override
+    protected boolean shouldExecuteOnFile(File file) {
+        Model model = PomHelper.getLocationAnnotatedModel(file);
+        String packaging = model.getPackaging();
+        if (POM.equals(packaging)) {
+            boolean profilesMatch = model.getProfiles()
+                    .stream()
```
Comment (carter-cundiff) [inline foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_8_0/Fabric8LocationMigration.java:50]: Will this be fine in the event there are no profiles in the pom? Will `.getProfiles()` return null?

Comment (ewilkins-csi) [reply foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_8_0/Fabric8LocationMigration.java:50]: `getProfiles` never returns null:
```java
public java.util.List<Profile> getProfiles() {
        if ( this.profiles == null )
        {
            this.profiles = new java.util.ArrayList<Profile>();
        }

        return this.profiles;
    }
```

**Review by csun-cpointe — APPROVED** [review]
LGTM

File: foundation/foundation-upgrade/src/test/resources/specifications/v1_8_0/orphedomos-to-fabric8-migration.feature (hunk 6 above)
Comment (ewilkins-csi) [reply foundation/foundation-upgrade/src/test/resources/specifications/v1_8_0/orphedomos-to-fabric8-migration.feature]: updated all the titles to be more clear

Comment (ewilkins-csi) [reply foundation/foundation-upgrade/src/test/resources/specifications/v1_8_0/orphedomos-to-fabric8-migration.feature]: updated all the titles to be more clear

**Review by carter-cundiff — APPROVED** [review]
